routing_common/bicycle_model.cpp

```diff
@@ -26,33 +26,33 @@ using namespace routing;
-    {HighwayType::HighwayTertiary, InOutCitySpeedKMpH(SpeedKMpH(15.0, 18.0), SpeedKMpH(20.0, 18.0))},
-    {HighwayType::HighwayTertiaryLink, InOutCitySpeedKMpH(SpeedKMpH(15.0, 18.0), SpeedKMpH(20.0, 18.0))},
-    {HighwayType::HighwayUnclassified, InOutCitySpeedKMpH(SpeedKMpH(12.0, 18.0))},
-
-    // https://github.com/organicmaps/organicmaps/issues/3881
-    // Set equal speeds here to avoid useless detours via service roads (Batumi_AvoidServiceDetour test).
-    {HighwayType::HighwayService, InOutCitySpeedKMpH(SpeedKMpH(10.0, 14.0))},
-    {HighwayType::HighwayResidential, InOutCitySpeedKMpH(SpeedKMpH(10.0, 14.0))},
-
-    {HighwayType::HighwayRoad, InOutCitySpeedKMpH(SpeedKMpH(10.0, 12.0))},
-    {HighwayType::HighwayTrack, InOutCitySpeedKMpH(SpeedKMpH(8.0, 12.0))},
-    {HighwayType::HighwayPath, InOutCitySpeedKMpH(SpeedKMpH(6.0, 12.0))},
-    {HighwayType::HighwayBridleway, InOutCitySpeedKMpH(SpeedKMpH(4.0, 12.0))},
-    {HighwayType::HighwayCycleway, InOutCitySpeedKMpH(SpeedKMpH(30.0, 20.0))},
-    {HighwayType::HighwayLivingStreet, InOutCitySpeedKMpH(SpeedKMpH(7.0, 8.0))},
+    // Note that roads with hwtag=yesbicycle get high speed of 0.9 * Cycleway.
+    /// @see Russia_UseTrunk test for Trunk weights.
+    {HighwayType::HighwayTrunk, InOutCitySpeedKMpH(SpeedKMpH(7.0, 17.0), SpeedKMpH(9.0, 19.0))},
+    {HighwayType::HighwayTrunkLink, InOutCitySpeedKMpH(SpeedKMpH(7.0, 17.0), SpeedKMpH(9.0, 19.0))},
+    {HighwayType::HighwayPrimary, InOutCitySpeedKMpH(SpeedKMpH(10.0, 17.0), SpeedKMpH(12.0, 19.0))},
+    {HighwayType::HighwayPrimaryLink, InOutCitySpeedKMpH(SpeedKMpH(10.0, 17.0), SpeedKMpH(12.0, 19.0))},
+    {HighwayType::HighwaySecondary, InOutCitySpeedKMpH(SpeedKMpH(13.0, 17.0), SpeedKMpH(15.0, 19.0))},
+    {HighwayType::HighwaySecondaryLink, InOutCitySpeedKMpH(SpeedKMpH(13.0, 17.0), SpeedKMpH(15.0, 19.0))},
+    {HighwayType::HighwayTertiary, InOutCitySpeedKMpH(SpeedKMpH(14.0, 17.0), SpeedKMpH(17.0, 19.0))},
+    {HighwayType::HighwayTertiaryLink, InOutCitySpeedKMpH(SpeedKMpH(14.0, 17.0), SpeedKMpH(17.0, 19.0))},
+    {HighwayType::HighwayUnclassified, InOutCitySpeedKMpH(SpeedKMpH(13.0, 17.0), SpeedKMpH(15.0, 19.0))},
+    {HighwayType::HighwayResidential, InOutCitySpeedKMpH(SpeedKMpH(12.0, 15.0), SpeedKMpH(14.0, 17.0))},
+    {HighwayType::HighwayService, InOutCitySpeedKMpH(SpeedKMpH(10.0, 12.0), SpeedKMpH(14.0, 17.0))},
+    {HighwayType::HighwayRoad, InOutCitySpeedKMpH(SpeedKMpH(10.0, 12.0), SpeedKMpH(14.0, 17.0))},
+
+    {HighwayType::HighwayTrack, InOutCitySpeedKMpH(SpeedKMpH(8.0, 12.0), SpeedKMpH(10.0, 14.0))},
+    {HighwayType::HighwayPath, InOutCitySpeedKMpH(SpeedKMpH(6.0, 10.0), SpeedKMpH(7.0, 12.0))},
+    {HighwayType::HighwayBridleway, InOutCitySpeedKMpH(SpeedKMpH(4.0, 10.0), SpeedKMpH(5.0, 12.0))},
+
+    {HighwayType::HighwayCycleway, InOutCitySpeedKMpH(SpeedKMpH(21.0, 18.0), SpeedKMpH(23.0, 20.0))},
+    {HighwayType::HighwayLivingStreet, InOutCitySpeedKMpH(SpeedKMpH(8.0, 8.0))},
```

Comment:
```suggestion
    {HighwayType::HighwayLivingStreet, InOutCitySpeedKMpH(SpeedKMpH(14.0, 10.0))},
```

this matches highway=tertiary in desirability and path/bridleway in speed

Reply:
or maybe give larger boost

I would say that after cycleway, good track, good service road, living street with asphalt surface is the next in order of preferred road types.

Reply:
What is your opinion on position of highway=residential, unclassified, tertiary amongst those?

Reply:
I've raised weights for living streets and service. But not as much. Logically a living_street isn't more desirable than a service road and we can't raise service much at the moment as it leads to many small useless detours "to visit a gas station" :)

We need some special logic to filter such small and pointless deviations from the main route.

Reply:
E.g. if the route has to go via a primary for several kms then there is no point to turn into gas stations driveways as 100m further one is back to the primary anyway.

Actually maybe introduction of turns/intersections penalties will help remedy this behavior.

Reply:
> Logically a living_street isn't more desirable than a service road and we can't raise service much at the moment as it leads to many small useless detours "to visit a gas station" :)

even if living_street is not better than service it may be still possible to achieve better routing by giving it cost lower than service

in general these are routing weigths, not intended to be used to compare cycling quality in consistent, transitive etc. fashion

> What is your opinion on position of highway=residential, unclassified, tertiary amongst those?

tertiary? definitely worse than residential and living_street, even for tertiary with bicycle lane due to higher traffic

though, tertiary with good asphalt may be better than service/residential/etc with terrible surface such as sand, mud, grass paver, true cobblestone and worse kinds of set

residential/unclassified is all over the place, but I would consider it (in my area) to be overall a tiny bit worse than living_street

Reply:
For some reason residentials had quite little weight before, maybe because is some areas its usual to have a lot of parked cars along? IDK.

I've made it higher but lower than tertiary still.

We can review it later again, but first would be good to find some test cases.

(here I talk about roads of similar quality only; OM applies a factor depending on the quality, see e.g. `paved_bad` in the file with weights)

Reply:
>  maybe because is some areas its usual to have a lot of parked cars along?

static parked cars are little to no problem (just do not cycle within doorzone, glued to side of the road)

maybe it was a performance optimization to prefer compute inferior route but faster? or attempt to prefer simpler instructions over busy roads?

> We can review it later again, but first would be good to find some test cases.

I will find some cases. I expect that no amount of weight tweaking will get match for all at once. That is OK.

prefer residential over tertiary (tertiary has high traffic and nasty curve with very poor visibility, and some local drivers taking shortcut over a wrong lane, I protested against using this hint too much but there is bicycle route going over residential. Residential has good surface and almost zero traffic.)

https://www.openstreetmap.org/directions?engine=graphhopper_bicycle&route=50.093257%2C19.814578%3B50.093415%2C19.807746#map=18/50.093343/19.811364&layers=N

![screen03](https://github.com/user-attachments/assets/7c61bbec-5adb-46a5-969f-3cec3d7a44e0)

----------------------------------

preferred route: https://www.openstreetmap.org/directions?engine=graphhopper_bicycle&route=50.095435%2C19.980408%3B50.098643%2C19.994554#map=16/50.09866/19.98585&layers=N

![screen05](https://github.com/user-attachments/assets/fc2dc35b-d983-42d4-a4ee-705b38ba61fc)

worse route: https://www.openstreetmap.org/directions?engine=graphhopper_bicycle&route=50.09480%2C19.97874%3B50.09864%2C19.99455 (no osm.org router passes, Google routes cyclists there)

![screen06](https://github.com/user-attachments/assets/f5987cf3-68b7-4afc-8100-1f508ca9acef)

though you need to dismount on a tiny bridge and has gravel section

but this tertiary has really heavy traffic

-----------------------------------------

bad https://www.openstreetmap.org/directions?engine=graphhopper_bicycle&route=50.078763%2C19.960291%3B50.078825%2C19.971396#map=17/50.077131/19.965656&layers=N

![screen07](https://github.com/user-attachments/assets/1091e340-687b-4733-8cfe-652e986d7b04)

good http://brouter.de/brouter-web/#map=17/50.07864/19.96423/standard&lonlats=19.971321,50.078749;19.962094,50.079217;19.960356,50.0789

![screen08](https://github.com/user-attachments/assets/c07f907e-0011-4798-84a7-a197ca8c266e)
though literally no router wanted to go there on its own

but https://www.openstreetmap.org/way/237026822 is nasty enough (lanes=4, cycleway=no, I added shoulder=no now) that going up/down on steps without ramp is still better

https://www.google.pl/maps/@50.0766128,19.9688628,3a,42.5y,216.52h,85.61t/data=!3m7!1e1!3m5!1sI9LitafSvcq_vcxbd4k1FA!2e0!6shttps:%2F%2Fstreetviewpixels-pa.googleapis.com%2Fv1%2Fthumbnail%3Fcb_client%3Dmaps_sv.tactile%26w%3D900%26h%3D600%26pitch%3D4.391837655803542%26panoid%3DI9LitafSvcq_vcxbd4k1FA%26yaw%3D216.5195172870467!7i16384!8i8192?entry=ttu&g_ep=EgoyMDI0MTExOS4yIKXMDSoASAFQAw%3D%3D

![screen09](https://github.com/user-attachments/assets/e7f781a8-59dc-4916-8fba-186e240c50f7)
have fun cycling here

-------------------------------------------------

http://brouter.de/brouter-web/#map=16/50.0756/19.9222/standard&lonlats=19.926238,50.07722;19.917955,50.078223

cycle for longer to avoid busy tertiary

![screen10](https://github.com/user-attachments/assets/3d36aebd-e9b7-4cda-92f8-aa8d4e8138a5)

------------------------------------------

prefers residential over tertiary + unsegregated footway allowing cyclists http://brouter.de/brouter-web/#map=16/50.0748/19.9204/standard&lonlats=19.921668,50.079108;19.916038,50.07401

![screen12](https://github.com/user-attachments/assets/f66b9e60-c74f-4127-ba60-d019896e8e0f)

Reply:
http://brouter.de/brouter-web/#map=15/50.0381/19.9428/standard&lonlats=19.945207,50.045255;19.94033,50.036663

prefers minor streets over busy tertiaries

![screen13](https://github.com/user-attachments/assets/32f09d05-9769-44cb-b392-29621cbec9d6)

Can add more, what kind of characteristics are not satisfied by tests so far?

Reply:
> prefer residential over tertiary (tertiary has high traffic and nasty curve with very poor visibility, and some local drivers taking shortcut over a wrong lane, I protested against using this hint too much but there is bicycle route going over residential. Residential has good surface and almost zero traffic.)
> 
> https://www.openstreetmap.org/directions?engine=graphhopper_bicycle&route=50.093257%2C19.814578%3B50.093415%2C19.807746
> 

OM passes. The residential has yesbicycle (prob inherited from route relation), so OM prefers it.

> 
> preferred route: https://www.openstreetmap.org/directions?engine=graphhopper_bicycle&route=50.095435%2C19.980408%3B50.098643%2C19.994554
> 

Passed.

> 
> worse route: https://www.openstreetmap.org/directions?engine=graphhopper_bicycle&route=50.09480%2C19.97874%3B50.09864%2C19.99455 (no osm.org router passes, Google routes cyclists there)
> 
> though you need to dismount on a tiny bridge and has gravel section
> 
> but this tertiary has really heavy traffic
> 

Not passed.

> bad https://www.openstreetmap.org/directions?engine=graphhopper_bicycle&route=50.078763%2C19.960291%3B50.078825%2C19.971396
> 
> 
> good http://brouter.de/brouter-web/#map=17/50.07864/19.96423/standard&lonlats=19.971321,50.078749;19.962094,50.079217;19.960356,50.0789
> 
> but https://www.openstreetmap.org/way/237026822 is nasty enough (lanes=4, cycleway=no, I added shoulder=no now) that going up/down on steps without ramp is still better
> especially that uphill part after going under railway bridge
> 

Not passed.


> 
> http://brouter.de/brouter-web/#map=16/50.0756/19.9222/standard&lonlats=19.926238,50.07722;19.917955,50.078223
> 
> cycle for longer to avoid busy tertiary - prefer to escape going slightly downhill (to East) than to go on it (to West)
> 
> again, this altitude may be not seen in OM data :(
> 

Not passed.

> 
> prefers residential over tertiary + unsegregated footway allowing cyclists http://brouter.de/brouter-web/#map=16/50.0748/19.9204/standard&lonlats=19.921668,50.079108;19.916038,50.07401

OM avoid the tertiary too, but in a different way

![image](https://github.com/user-attachments/assets/58b9af21-9917-46a0-bde7-3f1b9392bf2e)

Reply:
So let's leave for the next iteration.

Reply:
> OM avoid the tertiary too, but in a different way

this is not terrible and at this point it starts to get to "do you prefer to weave around pedestrians on crowded sidewalk or be on road with some cars" point, with preferred routes depending on person.

> OM passes

I would consider adding also some that pass now - this will protect against regressions

Reply:
- https://github.com/organicmaps/organicmaps/issues/9749